proto/cosmos/tx/v1beta1/tx.proto

```diff
@@ -116,6 +116,14 @@ message TxBody {
   // be processed by the chain
   uint64 timeout_height = 3;
 
+  // non_atomic is an atomicity flag. If false, the transaction will be atomic
+  // and be rejected if any of the messages fail. This is the default and
+  // compatible with previous versions of the SDK and TxBody.
+  //
+  // If set to true, the transaction will be executed as long as at least one
+  // of the messages succeeds.
+  bool non_atomic = 4;
```

Comment:
Transactions are definitionally atomic. A non-atomic transaction is nonsensical.

Reply:
The [original Tendermint paper](https://knowen-production.s3.amazonaws.com/uploads/attachment/file/1814/Buchman_Ethan_201606_Msater%2Bthesis.pdf) literally says that

> a transaction is an atomic operation ... [which] either completes or doesn’t occur at all, and can’t be left in an intermediate state

Reply:
Transactions being atomic is explicitly what we're trying to make more flexible here. They will still behave the same way when that parameter is set to false (the default), but will allow flexibility for users that need each message to be atomic within a transaction. The expected behaviour after the change is that sending a transaction with non_atomic=true will be equivalent to sending multiple sequential transactions to the mempool.

The execution guarantees are still present for each message and downstream assumptions can be adapted according to how the transaction is parametrised.

There are practical consideration as to why we need this. One particular case is relayers sending all the IBC messages they are relaying within the same transaction which means that one bad message would break the whole batch.

WRT the tendermint paper, that quote is not from the spec definition but from a description of the prior-art in the backround section, and comes from a reference of a paper that explicitly discusses the limitations of the transaction model and ways to overcome them. Regardless, specs need to be adaptable to their practical uses. The application layer introduced the concept of MultiMsg transactions (https://github.com/cosmos/cosmos-sdk/pull/1266) and users decided how to use them. I think updating the sdk to match those realities and make it easier for users is something to be desired.

Luckily, in an app-chain world, chains can just choose to to disallow non-atomic if that's their preference :)

Reply:
The tendermint guarantee is at a different layer of abstraction and definition of atomicity. @08d2 I believe your mixing up abstraction layers here. 

The Tendermint guarantee is at the definition of "Database atomicity".

A full tx is either included, or not included, by every single node.

The definition of atomicity here is different, as tge same data content is included by everything in the network, and being ran through the same function. The definition here is about whether the execution functions allows some messages to execute while some failing. Perhaps theres a better word than atomicity to reflect the abstraction difference.

Reply:
Consider an application that currently executes transactions without any knowledge of this new atomic flag. That application updates their SDK dependency to a new version which includes this proposed change. They re-build their application binary without making any changes that would accommodate this new flag. Their application will continue to compile successfully, right? (AFAICT, yes.) If so, will that node successfully join a network with peers that treat those messages with atomic=true differently? (AFAICT, yes, as this breaking change to the proto is not reflected in the proto version.) If so, will those two classes of nodes have different results when executing transactions marked as atomic=true? (AFAICT, yes.) If so, does that not break consensus? (AFAICT, yes?)

Reply:
> One particular case is relayers sending all the IBC messages they are relaying within the same transaction which means that one bad message would break the whole batch.

If a message needs to execute independently of other messages, then it can't be bundled into a transaction with those other messages. A relayer which bundles messages in this way is pretty unambiguously buggy.

Reply:
If a non-atomic tx contains a msg which is not correctly signed, will that block the overall tx, or will it allow other correctly-signed msgs to execute?

Reply:
wrt concensus, yeah, this probably needs to be a major semver change but old txs should still work as expected because the additive proto change defaults to false

> A relayer which bundles messages in this way is pretty unambiguously buggy.

I can't remember the exact reason they're doing this but I think its partly fees and partly avoiding account sequence mismatch issues. Might be worth also looking into getting them to change that though that's may be a social problem. Happy to move the solution up the stack if possible, though that would require going back and reevaluating the original requirements (IIRC the conclusion was that non-atomic multi-msg was the cleanest solution)

> If a non-atomic tx contains a msg which is not correctly signed, will that block the overall tx, or will it allow other correctly-signed msgs to execute?

This is a good question as signature validation may be happening before runMsgs. I can't context-switch right now, but I'll look into it when I can. If that does happen outside, the solutions are:
  * Define non-atomic as only applying to the message execution itself; or
  * Go over the other places where message validation happens and handle atomicity there too; that is signature validation checks (if they actually happen before) and ante handlers

Reply:
> > A relayer which bundles messages in this way is pretty unambiguously buggy.
>
> I can't remember the exact reason they're doing this but I think its partly fees and partly avoiding account sequence mismatch issues. Might be worth also looking into getting them to change that though that's may be a social problem. Happy to move the solution up the stack if possible, though that would require going back and reevaluating the original requirements (IIRC the conclusion was that non-atomic multi-msg was the cleanest solution)

I'd be curious to read any rationale behind that decision, when you find it.

Reply:
> Consider an application that currently executes transactions without any knowledge of this new atomic flag. That application updates their SDK dependency to a new version which includes this proposed change. They re-build their application binary without making any changes that would accommodate this new flag. Their application will continue to compile successfully, right? (AFAICT, yes.) If so, will that node successfully join a network with peers that treat those messages with atomic=true differently? (AFAICT, yes, as this breaking change to the proto is not reflected in the proto version.) If so, will those two classes of nodes have different results when executing the same transaction that's marked as atomic=true? (AFAICT, yes.) If so, does that not break consensus? (AFAICT, yes?)

You've just described any state breaking change? Every SDK minor release introduces state breaks?

I really think your having some abstraction errors here

Reply:
> Every SDK minor release introduces state breaks?

If that minor release produces results like those described above -- yes? Is that not clear?